File: src/lib/calendarSync.ts
import { tauriCommands } from './tauri';
import { isLikelyConnectivityError } from './vaultReplica';
import {
  normalizeCalendarDefinition,
  normalizeCalendarItem,
  type CalendarChangesPage,
  type CalendarDefinition,
  type CalendarLocation,
  type CalendarOperation,
  type CalendarRemoteChange,
  type CalendarSyncState,
} from '../types/calendar';

const CHANGE_PAGE_SIZE = 500;
const OPERATION_BATCH_SIZE = 500;
const MAX_CHANGE_PAGES_PER_PASS = 100;

export interface HostedCalendarOrigin {
  serverUrl: string;
  userId: string;
}

export type CalendarSyncPhase = 'discovering' | 'uploading' | 'downloading' | 'complete' | 'error';

export interface CalendarSyncProgress {
  originKey: string;
  serverUrl: string;
  phase: CalendarSyncPhase;
  processedItems: number;
  totalItems?: number;
}

export type CalendarSyncProgressListener = (progress: CalendarSyncProgress) => void;

export interface CalendarSyncAdapter {
  hostedCalendarRequest<T>(serverUrl: string, method: 'GET' | 'POST', path: string, body?: unknown): Promise<T>;
  calendarSave(profileId: string, calendar: CalendarDefinition): Promise<void>;
  calendarReadSyncState(profileId: string, originKey: string): Promise<CalendarSyncState | null>;
  calendarWriteSyncState(profileId: string, state: CalendarSyncState): Promise<void>;
  calendarListPendingOperations(profileId: string): Promise<CalendarOperation[]>;
  calendarAcknowledgeOperations(profileId: string, clientOperationIds: string[]): Promise<void>;
  calendarMarkOperationFailed(profileId: string, clientOperationId: string, error: string, attemptedAt: string): Promise<void>;
  calendarApplyRemoteChanges(profileId: string, changes: CalendarRemoteChange[], state: CalendarSyncState): Promise<void>;
}

export interface CalendarOriginSyncResult {
  originKey: string;
  serverUrl: string;
  appliedChanges: number;
  replayedOperations: number;
  failedOperations: number;
  completedAt: string;
  error?: string;
}

export function hostedCalendarOriginKey(origin: HostedCalendarOrigin): string {
  return `${origin.serverUrl.replace(/\/$/, '')}::${origin.userId}`;
}

export function normalizeHostedCalendarOrigins(
  origins: HostedCalendarOrigin[],
): HostedCalendarOrigin[] {
  return Array.from(new Map(origins.map((origin) => {
    const normalized = { ...origin, serverUrl: origin.serverUrl.replace(/\/$/, '') };
    return [hostedCalendarOriginKey(normalized), normalized];
  })).values()).sort((left, right) => (
    hostedCalendarOriginKey(left).localeCompare(hostedCalendarOriginKey(right))
  ));
}

function operationCalendarId(operation: CalendarOperation): string | undefined {
  switch (operation.mutation.type) {
    case 'createCalendar':
    case 'updateCalendar':
      return operation.mutation.calendar.id;
    case 'deleteCalendar':
    case 'deleteItem':
      return operation.mutation.calendarId;
    case 'upsertItem':
      return operation.mutation.item.calendarId;
  }
}

function chunks<T>(values: T[], size: number): T[][] {
  const result: T[][] = [];
  for (let index = 0; index < values.length; index += size) {
    result.push(values.slice(index, index + size));
  }
  return result;
}

function errorMessage(error: unknown): string {
  return (error instanceof Error ? error.message : String(error)).slice(0, 2_000);
}

async function replayOperationBatch(
  profileId: string,
  serverUrl: string,
  operations: CalendarOperation[],
  adapter: CalendarSyncAdapter,
  onSettled: (count: number) => void,
): Promise<{ replayedOperations: number; failedOperations: number }> {
  if (operations.length === 0) return { replayedOperations: 0, failedOperations: 0 };
  try {
    await adapter.hostedCalendarRequest(
      serverUrl,
      'POST',
      '/api/v1/calendars/operations',
      { operations },
    );
    await adapter.calendarAcknowledgeOperations(
      profileId,
      operations.map((operation) => operation.clientOperationId),
    );
    onSettled(operations.length);
    return { replayedOperations: operations.length, failedOperations: 0 };
  } catch (error) {
    if (isLikelyConnectivityError(error)) throw error;
    if (operations.length > 1) {
      const middle = Math.ceil(operations.length / 2);
      const first = await replayOperationBatch(profileId, serverUrl, operations.slice(0, middle), adapter, onSettled);
      const second = await replayOperationBatch(profileId, serverUrl, operations.slice(middle), adapter, onSettled);
      return {
        replayedOperations: first.replayedOperations + second.replayedOperations,
        failedOperations: first.failedOperations + second.failedOperations,
      };
    }
    await adapter.calendarMarkOperationFailed(
      profileId,
      operations[0].clientOperationId,
      errorMessage(error),
      new Date().toISOString(),
    );
    onSettled(1);
    return { replayedOperations: 0, failedOperations: 1 };
  }
}

function hostedLocation(origin: HostedCalendarOrigin): CalendarLocation {
  return { kind: 'hosted', serverUrl: origin.serverUrl, userId: origin.userId };
}

function normalizeRemoteCalendar(value: unknown, origin: HostedCalendarOrigin): CalendarDefinition {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    throw new Error('The server returned an invalid calendar definition.');
  }
  const location = (value as { location?: unknown }).location;
  if (location && typeof location === 'object' && !Array.isArray(location)
    && (location as { kind?: unknown }).kind === 'kanban') {
    const originKey = (location as { originKey?: unknown }).originKey;
    if (typeof originKey !== 'string' || !originKey.trim()) {
      throw new Error('The server returned an invalid Kanban calendar origin.');
    }
    return normalizeCalendarDefinition({
      ...value,
      location: {
        kind: 'kanban',
        originKey: `${origin.serverUrl.replace(/\/$/, '')}::${originKey}`,
      },
      readOnly: true,
    });
  }
  return normalizeCalendarDefinition({ ...value, location: hostedLocation(origin) });
}

function normalizeRemoteChange(
  change: CalendarRemoteChange,
  origin: HostedCalendarOrigin,
): CalendarRemoteChange {
  if (change.operation !== 'upsert') return change;
  if (change.entityType === 'calendar') {
    return { ...change, payload: normalizeRemoteCalendar(change.payload, origin) };
  }
  const item = normalizeCalendarItem(change.payload);
  if (item.sourceBinding?.kind !== 'kanban') return { ...change, payload: item };
  return {
    ...change,
    payload: {
      ...item,
      sourceBinding: {
        ...item.sourceBinding,
        serverUrl: origin.serverUrl.replace(/\/$/, ''),
      },
    },
  };
}

async function discoverCalendars(
  profileId: string,
  origin: HostedCalendarOrigin,
  adapter: CalendarSyncAdapter,
): Promise<CalendarDefinition[]> {
  const values = await adapter.hostedCalendarRequest<unknown[]>(
    origin.serverUrl,
    'GET',
    '/api/v1/calendars',
  );
  const calendars = values.map((value) => normalizeRemoteCalendar(value, origin));
  for (const calendar of calendars) {
    await adapter.calendarSave(profileId, calendar);
  }
  return calendars;
}

export async function syncHostedCalendarOrigin(
  profileId: string,
  origin: HostedCalendarOrigin,
  cachedCalendars: CalendarDefinition[],
  onProgress?: CalendarSyncProgressListener,
  adapter: CalendarSyncAdapter = tauriCommands,
): Promise<CalendarOriginSyncResult> {
  const originKey = hostedCalendarOriginKey(origin);
  const report = (phase: CalendarSyncPhase, processedItems: number, totalItems?: number) => {
    onProgress?.({ originKey, serverUrl: origin.serverUrl, phase, processedItems, totalItems });
  };
  report('discovering', 0);
  let cursor = 0;
  let appliedChanges = 0;
  let replayedOperations = 0;
  let failedOperations = 0;
  try {
    cursor = Number.parseInt(
      (await adapter.calendarReadSyncState(profileId, originKey))?.cursor ?? '0',
      10,
    );
    if (!Number.isSafeInteger(cursor) || cursor < 0) cursor = 0;
    const discovered = await discoverCalendars(profileId, origin, adapter);
    const calendarIds = new Set([
      ...discovered.map((calendar) => calendar.id),
      ...cachedCalendars
        .filter((calendar) => calendar.location.kind === 'hosted'
          && hostedCalendarOriginKey(calendar.location) === originKey)
        .map((calendar) => calendar.id),
    ]);
    const pending = (await adapter.calendarListPendingOperations(profileId))
      .filter((operation) => {
        const calendarId = operationCalendarId(operation);
        return calendarId != null && calendarIds.has(calendarId);
      });
    let settledOperations = 0;
    report('uploading', 0, pending.length);
    for (const batch of chunks(pending, OPERATION_BATCH_SIZE)) {
      const replay = await replayOperationBatch(profileId, origin.serverUrl, batch, adapter, (count) => {
        settledOperations += count;
        report('uploading', settledOperations, pending.length);
      });
      replayedOperations += replay.replayedOperations;
      failedOperations += replay.failedOperations;
    }

    if (failedOperations > 0) {
      const message = `${failedOperations} calendar change${failedOperations === 1 ? '' : 's'} need attention.`;
      report('error', settledOperations, pending.length);
      await adapter.calendarWriteSyncState(profileId, {
        originKey,
        cursor: String(cursor),
        lastError: message,
      });
      return {
        originKey,
        serverUrl: origin.serverUrl,
        appliedChanges,
        replayedOperations,
        failedOperations,
        completedAt: new Date().toISOString(),
        error: message,
      };
    }

    report('downloading', 0);
    for (let pageIndex = 0; pageIndex < MAX_CHANGE_PAGES_PER_PASS; pageIndex += 1) {
      const page = await adapter.hostedCalendarRequest<CalendarChangesPage>(
        origin.serverUrl,
        'GET',
        `/api/v1/calendars/changes?cursor=${cursor}&limit=${CHANGE_PAGE_SIZE}`,
      );
      if (!Number.isSafeInteger(page.cursor) || page.cursor < cursor) {
        throw new Error('The server returned an invalid calendar change cursor.');
      }
      if (page.hasMore && (page.changes.length === 0 || page.cursor === cursor)) {
        throw new Error('The server returned a non-advancing calendar change page.');
      }
      const changes = page.changes.map((change) => normalizeRemoteChange(change, origin));
      cursor = page.cursor;
      const state: CalendarSyncState = {
        originKey,
        cursor: String(cursor),
        lastSyncedAt: new Date().toISOString(),
      };
      await adapter.calendarApplyRemoteChanges(profileId, changes, state);
      appliedChanges += changes.length;
      report('downloading', appliedChanges);
      if (!page.hasMore) break;
      if (pageIndex === MAX_CHANGE_PAGES_PER_PASS - 1) {
        throw new Error('Calendar sync reached its bounded page limit.');
      }
    }
    report('complete', replayedOperations + appliedChanges, replayedOperations + appliedChanges);
    return {
      originKey,
      serverUrl: origin.serverUrl,
      appliedChanges,
      replayedOperations,
      failedOperations,
      completedAt: new Date().toISOString(),
    };
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    report('error', replayedOperations + appliedChanges);
    await adapter.calendarWriteSyncState(profileId, {
      originKey,
      cursor: String(cursor),
      lastError: message,
    }).catch(() => {});
    return {
      originKey,
      serverUrl: origin.serverUrl,
      appliedChanges,
      replayedOperations,
      failedOperations,
      completedAt: new Date().toISOString(),
      error: message,
    };
  }
}

export async function syncHostedCalendars(
  profileId: string,
  origins: HostedCalendarOrigin[],
  cachedCalendars: CalendarDefinition[],
  onProgress?: CalendarSyncProgressListener,
  adapter: CalendarSyncAdapter = tauriCommands,
): Promise<CalendarOriginSyncResult[]> {
  const uniqueOrigins = normalizeHostedCalendarOrigins(origins);
  return Promise.all(uniqueOrigins.map((origin) => (
    syncHostedCalendarOrigin(profileId, origin, cachedCalendars, onProgress, adapter)
  )));
}

File: src-tauri/src/background/calendar_sync.rs
use super::sync::{
    classify_session_error, ensure_running, request_json, JobExecutionError, JobExecutionSummary,
};
use super::{BackgroundCoordinator, BackgroundJobProgress, BackgroundJobRequest};
use crate::hosted_session::refresh_session_locked;
use collab_calendar::{
    CalendarDefinition, CalendarLocation, CalendarMutation, CalendarOperation,
    CalendarRemoteChange, CalendarSyncState,
};
use reqwest::Method;
use serde::Deserialize;
use serde_json::{json, Value};
use std::collections::HashSet;
use std::sync::atomic::AtomicBool;
use std::time::{Duration, Instant};

const CHANGE_PAGE_SIZE: u32 = 500;
const OPERATION_BATCH_SIZE: usize = 500;
const MAX_CHANGE_PAGES_PER_PASS: usize = 100;

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct CalendarChangesPage {
    cursor: i64,
    changes: Vec<CalendarRemoteChange>,
    has_more: bool,
}

pub(crate) async fn run_calendar_sync(
    coordinator: &BackgroundCoordinator,
    job_id: &str,
    request: &BackgroundJobRequest,
    cancel: &AtomicBool,
    budget: Duration,
) -> Result<JobExecutionSummary, JobExecutionError> {
    let started = Instant::now();
    let server_url = request
        .server_url
        .as_deref()
        .ok_or_else(|| JobExecutionError::persistence("Calendar sync requires a server URL."))?;
    let profile_id = request
        .profile_id
        .as_deref()
        .ok_or_else(|| JobExecutionError::persistence("Calendar sync requires a profile ID."))?;
    let registration = coordinator
        .list_servers()
        .map_err(JobExecutionError::persistence)?
        .into_iter()
        .find(|entry| entry.server_url == server_url)
        .ok_or_else(|| {
            JobExecutionError::interrupted(
                "The server is not registered for native background work.",
            )
        })?;
    if !registration.background_sync_enabled {
        return Err(JobExecutionError::interrupted(
            "Background synchronization is disabled for this server.",
        ));
    }
    ensure_running(cancel, started, budget)?;
    let session = refresh_session_locked(
        &coordinator.sessions,
        server_url,
        registration.allow_invalid_certificates,
        registration.persist_across_reboots,
        true,
    )
    .await
    .map_err(classify_session_error)?;
    let user_id = session.user.id.clone();
    let origin_key = format!("{}::{user_id}", session.server_url);
    let store = crate::commands::calendar::store(profile_id)
        .await
        .map_err(JobExecutionError::persistence)?;
    let old_state = store
        .read_sync_state(&origin_key)
        .await
        .map_err(|error| JobExecutionError::persistence(error.to_string()))?;
    let mut cursor = old_state
        .as_ref()
        .and_then(|state| state.cursor.as_deref())
        .and_then(|cursor| cursor.parse::<i64>().ok())
        .filter(|cursor| *cursor >= 0)
        .unwrap_or(0);

    coordinator
        .update_progress(
            job_id,
            BackgroundJobProgress {
                completed: 0,
                total: None,
                detail: Some("Discovering hosted calendars".to_string()),
            },
        )
        .map_err(JobExecutionError::persistence)?;
    let remote_calendars: Vec<Value> =
        request_json(&session, Method::GET, "/api/v1/calendars", None).await?;
    let mut calendar_ids = HashSet::new();
    for value in remote_calendars {
        ensure_running(cancel, started, budget)?;
        let calendar = normalize_remote_calendar(value, &session.server_url, &user_id)?;
        calendar_ids.insert(calendar.id.clone());
        store
            .upsert_calendar(&calendar)
            .await
            .map_err(|error| JobExecutionError::persistence(error.to_string()))?;
    }
    for calendar in store
        .list_calendars()
        .await
        .map_err(|error| JobExecutionError::persistence(error.to_string()))?
    {
        if matches!(
            calendar.location,
            CalendarLocation::Hosted {
                ref server_url,
                ref user_id
            } if server_url.trim_end_matches('/') == session.server_url && user_id == &session.user.id
        ) {
            calendar_ids.insert(calendar.id);
        }
    }

    let pending = store
        .list_pending_operations()
        .await
        .map_err(|error| JobExecutionError::persistence(error.to_string()))?
        .into_iter()
        .filter(|operation| {
            operation_calendar_id(operation)
                .is_some_and(|calendar_id| calendar_ids.contains(calendar_id))
        })
        .collect::<Vec<_>>();
    let mut replayed = 0_u64;
    let mut failed = 0_u64;
    for batch in pending.chunks(OPERATION_BATCH_SIZE) {
        ensure_running(cancel, started, budget)?;
        coordinator
            .update_progress(
                job_id,
                BackgroundJobProgress {
                    completed: replayed + failed,
                    total: Some(pending.len() as u64),
                    detail: Some("Uploading calendar changes".to_string()),
                },
            )
            .map_err(JobExecutionError::persistence)?;
        let result = replay_operation_groups(&session, &store, batch).await?;
        replayed += result.0;
        failed += result.1;
    }
    if failed > 0 {
        store
            .write_sync_state(&CalendarSyncState {
                origin_key,
                cursor: Some(cursor.to_string()),
                last_synced_at: old_state.and_then(|state| state.last_synced_at),
                last_error: Some(format!("{failed} calendar change(s) require attention.")),
            })
            .await
            .map_err(|error| JobExecutionError::persistence(error.to_string()))?;
        return Ok(JobExecutionSummary {
            completed: replayed + failed,
            total: replayed + failed,
            failed,
            message: format!("{failed} calendar change(s) require attention"),
        });
    }

    let mut applied = 0_u64;
    for page_index in 0..MAX_CHANGE_PAGES_PER_PASS {
        ensure_running(cancel, started, budget)?;
        coordinator
            .update_progress(
                job_id,
                BackgroundJobProgress {
                    completed: applied,
                    total: None,
                    detail: Some("Downloading calendar changes".to_string()),
                },
            )
            .map_err(JobExecutionError::persistence)?;
        let path = format!("/api/v1/calendars/changes?cursor={cursor}&limit={CHANGE_PAGE_SIZE}");
        let page: CalendarChangesPage = request_json(&session, Method::GET, &path, None).await?;
        if page.cursor < cursor
            || (page.has_more && (page.changes.is_empty() || page.cursor == cursor))
        {
            return Err(JobExecutionError::persistence(
                "The server returned a non-advancing calendar change page.",
            ));
        }
        let changes = page
            .changes
            .into_iter()
            .map(|change| normalize_remote_change(change, &session.server_url, &user_id))
            .collect::<Result<Vec<_>, _>>()?;
        cursor = page.cursor;
        let state = CalendarSyncState {
            origin_key: origin_key.clone(),
            cursor: Some(cursor.to_string()),
            last_synced_at: Some(chrono::Utc::now().to_rfc3339()),
            last_error: None,
        };
        store
            .apply_remote_changes(&changes, &state)
            .await
            .map_err(|error| JobExecutionError::persistence(error.to_string()))?;
        applied += changes.len() as u64;
        if !page.has_more {
            break;
        }
        if page_index == MAX_CHANGE_PAGES_PER_PASS - 1 {
            return Err(JobExecutionError::interrupted(
                "Calendar sync reached its bounded page limit.",
            ));
        }
    }

    Ok(JobExecutionSummary {
        completed: replayed + failed + applied,
        total: replayed + failed + applied,
        failed,
        message: format!("Replayed {replayed} and applied {applied} calendar change(s)"),
    })
}

async fn replay_batch(
    session: &crate::state::ServerSessionState,
    operations: &[CalendarOperation],
) -> Result<(), JobExecutionError> {
    let _: Value = request_json(
        session,
        Method::POST,
        "/api/v1/calendars/operations",
        Some(json!({ "operations": operations })),
    )
    .await?;
    Ok(())
}

async fn replay_operation_groups(
    session: &crate::state::ServerSessionState,
    store: &collab_calendar::CalendarStore,
    operations: &[CalendarOperation],
) -> Result<(u64, u64), JobExecutionError> {
    let mut groups = vec![operations.to_vec()];
    let mut replayed = 0;
    let mut failed = 0;
    while let Some(group) = groups.pop() {
        match replay_batch(session, &group).await {
            Ok(()) => {
                let ids = group
                    .iter()
                    .map(|operation| operation.client_operation_id.clone())
                    .collect::<Vec<_>>();
                store
                    .acknowledge_operations(&ids)
                    .await
                    .map_err(|error| JobExecutionError::persistence(error.to_string()))?;
                replayed += group.len() as u64;
            }
            Err(error) if error.retryable => return Err(error),
            Err(_) if group.len() > 1 => {
                let middle = group.len().div_ceil(2);
                groups.push(group[middle..].to_vec());
                groups.push(group[..middle].to_vec());
            }
            Err(error) => {
                let operation = &group[0];
                store
                    .mark_operation_failed(
                        &operation.client_operation_id,
                        &error.message,
                        &chrono::Utc::now().to_rfc3339(),
                    )
                    .await
                    .map_err(|store_error| {
                        JobExecutionError::persistence(store_error.to_string())
                    })?;
                failed += 1;
            }
        }
    }
    Ok((replayed, failed))
}

fn operation_calendar_id(operation: &CalendarOperation) -> Option<&str> {
    match &operation.mutation {
        CalendarMutation::CreateCalendar { calendar }
        | CalendarMutation::UpdateCalendar { calendar } => Some(&calendar.id),
        CalendarMutation::DeleteCalendar { calendar_id }
        | CalendarMutation::DeleteItem { calendar_id, .. } => Some(calendar_id),
        CalendarMutation::UpsertItem { item } => Some(&item.calendar_id),
    }
}

fn normalize_remote_calendar(
    mut value: Value,
    server_url: &str,
    user_id: &str,
) -> Result<CalendarDefinition, JobExecutionError> {
    let location_kind = value
        .get("location")
        .and_then(|location| location.get("kind"))
        .and_then(Value::as_str);
    let location = match location_kind {
        Some("kanban") => {
            let origin_key = value
                .get("location")
                .and_then(|location| location.get("originKey"))
                .and_then(Value::as_str)
                .map(str::to_string)
                .ok_or_else(|| {
                    JobExecutionError::persistence(
                        "The server returned an invalid Kanban calendar origin.",
                    )
                })?;
            set_read_only(&mut value);
            json!({ "kind": "kanban", "originKey": format!("{server_url}::{origin_key}") })
        }
        Some("subscription") => {
            let subscription_id = value
                .get("location")
                .and_then(|location| location.get("subscriptionId"))
                .and_then(Value::as_str)
                .map(str::to_string)
                .ok_or_else(|| {
                    JobExecutionError::persistence(
                        "The server returned an invalid calendar subscription.",
                    )
                })?;
            set_read_only(&mut value);
            json!({
                "kind": "subscription",
                "subscriptionId": subscription_id,
                "serverUrl": server_url,
                "userId": user_id
            })
        }
        _ => json!({ "kind": "hosted", "serverUrl": server_url, "userId": user_id }),
    };
    value
        .as_object_mut()
        .ok_or_else(|| {
            JobExecutionError::persistence("The server returned an invalid calendar definition.")
        })?
        .insert("location".to_string(), location);
    serde_json::from_value(value).map_err(|_| {
        JobExecutionError::persistence("The server returned an invalid calendar definition.")
    })
}

fn normalize_remote_change(
    mut change: CalendarRemoteChange,
    server_url: &str,
    user_id: &str,
) -> Result<CalendarRemoteChange, JobExecutionError> {
    if change.operation != "upsert" {
        return Ok(change);
    }
    if change.entity_type == "calendar" {
        let calendar = normalize_remote_calendar(
            change.payload.take().ok_or_else(|| {
                JobExecutionError::persistence("Remote calendar upsert has no payload.")
            })?,
            server_url,
            user_id,
        )?;
        change.payload = Some(
            serde_json::to_value(calendar)
                .map_err(|error| JobExecutionError::persistence(error.to_string()))?,
        );
    } else if change.entity_type == "item" {
        if let Some(payload) = change.payload.as_mut() {
            if payload
                .get("sourceBinding")
                .and_then(|binding| binding.get("kind"))
                .and_then(Value::as_str)
                == Some("kanban")
            {
                if let Some(binding) = payload
                    .get_mut("sourceBinding")
                    .and_then(Value::as_object_mut)
                {
                    binding.insert(
                        "serverUrl".to_string(),
                        Value::String(server_url.to_string()),
                    );
                }
            }
        }
    }
    Ok(change)
}

fn set_read_only(value: &mut Value) {
    if let Some(object) = value.as_object_mut() {
        object.insert("readOnly".to_string(), Value::Bool(true));
    }
}
